client-dynamodb-v2/src/main/kotlin/app/cash/backfila/client/dynamodbv2/internal/DynamoDbBackfillOperator.kt

```diff
@@ -0,0 +1,171 @@
+package app.cash.backfila.client.dynamodbv2.internal
+
+import app.cash.backfila.client.dynamodbv2.DynamoDbBackfill
+import app.cash.backfila.client.spi.BackfilaParametersOperator
+import app.cash.backfila.client.spi.BackfillOperator
+import app.cash.backfila.protos.clientservice.GetNextBatchRangeRequest
+import app.cash.backfila.protos.clientservice.GetNextBatchRangeResponse
+import app.cash.backfila.protos.clientservice.KeyRange
+import app.cash.backfila.protos.clientservice.PrepareBackfillRequest
+import app.cash.backfila.protos.clientservice.PrepareBackfillResponse
+import app.cash.backfila.protos.clientservice.RunBatchRequest
+import app.cash.backfila.protos.clientservice.RunBatchResponse
+import com.google.common.base.Stopwatch
+import software.amazon.awssdk.services.dynamodb.DynamoDbClient
+import software.amazon.awssdk.services.dynamodb.model.AttributeValue
+import software.amazon.awssdk.services.dynamodb.model.BillingMode
+import software.amazon.awssdk.services.dynamodb.model.ScanRequest
+import java.time.Duration
+
+class DynamoDbBackfillOperator<I : Any, P : Any>(
+  private val dynamoDbClient: DynamoDbClient,
+  override val backfill: DynamoDbBackfill<I, P>,
+  private val parametersOperator: BackfilaParametersOperator<P>,
+  private val keyRangeCodec: DynamoDbKeyRangeCodec,
+) : BackfillOperator {
+  override fun name(): String = backfill.javaClass.toString()
+
+  override fun prepareBackfill(request: PrepareBackfillRequest): PrepareBackfillResponse {
+    val config =
+      parametersOperator.constructBackfillConfig(request.parameters, request.dry_run)
+    backfill.validate(config)
+
+    if (backfill.mustHaveProvisionedBillingMode()) {
+      var tableResponse = dynamoDbClient.describeTable {
+        it.tableName(backfill.dynamoDbTable.tableName())
+      }
+      var billingModeSummary = tableResponse.table().billingModeSummary()
+      // It's odd but a null billingModeSummary implies "PROVISIONED"
+      require(
+        billingModeSummary == null ||
+          billingModeSummary.billingMode() == BillingMode.PROVISIONED
+      ) {
+        "Trying to prepare a backfill on a Dynamo table named ${backfill.dynamoDbTable.tableName()} " +
+          "with a billing mode that is not PROVISIONED, it is " +
+          "${backfill.dynamoDbTable.tableName()}. This can get very expensive. " +
+          "Please provision your dynamo capacity for this table and try again."
+      }
+    }
+
+    // TODO(mikepaw): dynamically select the segment count by probing DynamoDB.
+    val segmentCount = backfill.fixedSegmentCount(config) ?: 2048
+    val partitionCount = backfill.partitionCount(config)
```

Comment:
v1 client uses largest power of 2 < 1,000,000 for this value.

```suggestion
    // TODO(mikepaw): dynamically select the segment count by probing DynamoDB.
    // The maximum number of segments is 1,000,000 so we are currently setting it to the largest power 
    // of 2 less than this value.
    val segmentCount = backfill.fixedSegmentCount(config) ?: 524288
    val partitionCount = backfill.partitionCount(config)
```